File: algo/inversion_num/inversion.cpp
#include <iostream>
using namespace std;
 
long mergeSort(long arr[], long temp[], long left, long right);
long merge(long arr[], long temp[], long left, long mid, long right);

long mergeSort(long arr[], long temp[], long left, long right)
{
    long mid, inv = 0;
    if (right > left) {

        mid = (right + left) / 2;
        inv += mergeSort(arr, temp, left, mid);
        inv += mergeSort(arr, temp, mid + 1, right);
        inv += merge(arr, temp, left, mid + 1, right);
    }
    return inv % 524287;
}
 

long merge(long arr[], long temp[], long left, long mid, long right) {
    long i, j, k;
    long inv = 0;
 
    i = left;
    j = mid;
    k = left;
    while ((i <= mid - 1) && (j <= right)) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        }
        else {
            temp[k++] = arr[j++];
            inv = inv + (mid - i);
        }
    }
 
    while (i <= mid - 1) temp[k++] = arr[i++];
    while (j <= right) temp[k++] = arr[j++];
    for (i = left; i <= right; i++) arr[i] = temp[i];
 
    return inv ;
}
 

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    long N;
    long input;
    cin >> N;
    long arr[N];
    long temp[N];
    for (long i=0; i<N; i++){
        cin >> input;
        arr[i] = input;
    }
    
    long ans = mergeSort(arr, temp, 0, N-1);
    cout << ans;

    return 0;
}

File: codes/ncku/inversion_num/inversion.py
def mergeSortInversions(arr):
    if len(arr) == 1:
        return arr, 0
    else:
        a = arr[:len(arr)//2]
        b = arr[len(arr)//2:]
        a, ai = mergeSortInversions(a)
        b, bi = mergeSortInversions(b)
        c = []
        i = 0
        j = 0
        inv = 0 + ai + bi

        while i < len(a) and j < len(b):
            if int(a[i]) <= int(b[j]):
                c.append(a[i])
                i += 1
            else:
                c.append(b[j])
                j += 1
                inv += (len(a)-i)
        c += a[i:]
        c += b[j:]
        return c, inv


N = input()
arr = input().split(" ")


# print(mergeSortInversions(arr)[1])